knowledge-base/adr/0038-csrf-middleware.md

```diff
@@ -0,0 +1,106 @@
+    );
+```
+
+Currently, the middleware is only added to the `executeFn()` to the target system.
+For the fetch request it is not considered.
+Note that any kind of resilience entering via the middleware is not considered.
+
+### General Improvement - Disable Fetch CSRF by Default
+
+Not all systems require a CSRF token.
+We follow also the trend of optional functionality not being enabled by default, like cache.
+Currently, the CSRF token fails gracefully, and the SDK retries to get one, this means we may have unnecessary HTTP requests.
+If the default is off, one should really fail.
+
+## CSRF as one Middleware Layer
+
+Do the CSRF token fetching in a middleware.
+In version 3 the `csrf()` middleware is added per default so this illustrates just the
+
+```ts
+executeWithMiddleware([csrf(), timeout(), circuitBreaker(), retry()], {
+  context,
+  fn: axiosRequest,
+  fnArgument: axiosPostRequestConfit
+});
+```
+
+The `csrf()` middleware will fetch the token and add it to the `fn` before it is executed:
+
+```ts
+function csrf() {
+  return (
+    fn: Function<HttpResponse, HttpRequest>,
+    context: HttpRequestContext
+  ) => {
+    return async () => {
+      const csrfToken = await getCsrfToken(context);
+      fn.arg.headers['x-csrf-token'] = csrfToken;
+      return fn();
+    };
+  };
+}
+```
+
+Note that the order of middlewares is crucial.
+Middlewares are added from left to right to `fn()` and execution is then from right to left.
+The initial function `fn()` is added for illustration here:
+
+```ts
+// retry around everything
+// timeout for fn and csrf together
+// cb blocking fn and csrf
+executeWithMiddleware([fn(), csrf(), timeout(), circuitBreaker(), retry()]);
+
+// like above but timeout only for fn(), no timeout for the csrf call
+executeWithMiddleware([fn(), timeout(), csrf(), circuitBreaker(), retry()]);
+
+// csrf() completely outside the resilience
+executeWithMiddleware([fn(), timeout(), circuitBreaker(), retry(), csrf()]);
+```
+
+Note that the csrf requests are not recorded in the circuit breaker history.
+A dedicated timeout or retry is not possible as well.
+To allow this you would need to pass a dedicated middleware to it `csrf([timeout(),circuitBreaker()])`
+We could offer a default middleware including `csrf()` which does nothing for `GET` requests:
+
+```ts
+const defaultMiddleware = ([
+  timeout(),
+  circuitBreaker(),
+  csrf([timeout(), circuitBreaker()])
+] = [...resilience(), csrf(resilience())]);
+```
+
+This means:
+
+- dedicated timeout for the actual request and csrf fetching
+- circuit breaker instance based on destination URL will record actual request and csrf fetching
+- retry is not enabled per default
+
+In case a user need something different they can pass their own `csrf()` middleware to adjust the method or `URL`.
+
+## Migration Discussion
+
+- Too late in version 3 development to change the CSRF default behavior.
+- Deprecate `skipCsrfTokenFetchion()` in version 3 and remove it in version 4.
```

Comment:
[req] I assume this is meant: https://sap.github.io/cloud-sdk/api/v2/interfaces/sap_cloud_sdk_http_client.HttpRequestOptions.html#fetchCsrfToken
```suggestion
- Deprecate `fetchCsrfToken` in version 3 and remove it in version 4.
```

Reply:
There are two optins: `skipCsrfTokenFetching` on the typed clients and `fetchCsrfToken` on the http-client.

Reply:
I removed the typo and included them both.